File: Mar-2026/02-AddTwoNumbers.py
from typing import Optional


# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def addTwoNumbers(
        self, l1: Optional[ListNode], l2: Optional[ListNode]
    ) -> Optional[ListNode]:
        if l1:
            if l2:
                # if l1 and l2 are not empty
                l3 = ListNode(l1.val + l2.val)
            else:
                # if l1 is not empty but l2 is empty
                l3 = ListNode(l1.val)
            
            if l3.val < 10:
                l1_next = l1.next
            else:
                l3.val -= 10
                # make sure it won't change original l1
                l1_next = (
                    ListNode(l1.next.val + 1, l1.next.next) # works
                    if type(l1.next) is ListNode
                    else ListNode(1)
                ) 
                
        elif l2:
            # if l1 is empty but l2 is not empty
            l3 = ListNode(l2.val)
            l1_next = None
        else:
            # both l1 and l2 are None
            return None

        l3.next = self.addTwoNumbers(l1_next, l2.next if l2 else None)
        return l3

def show_ll(ll: Optional[ListNode]) -> None:
    if not ll:
        print("None")
    else:
        while type(ll.next) is ListNode:
            print(ll.val, end="")
            ll = ll.next
        print(ll.val)


if __name__ == "__main__":
    demo = Solution()
    l1 = ListNode(2, ListNode(4, ListNode(3)))
    l2 = ListNode(5, ListNode(6, ListNode(4)))
    l3 = demo.addTwoNumbers(l1, l2)
    show_ll(l3)
    
    l1 = ListNode(4)
    l2 = ListNode(6)
    l3 = demo.addTwoNumbers(l1, l2)
    show_ll(l3)

File: Mar-2026/02-AddTwoNumbers.cc
#include <stdio.h>

/* Definition for singly-linked list. */
struct ListNode {
  int val;
  ListNode *next;
  ListNode() : val(0), next(nullptr) {}
  ListNode(int x) : val(x), next(nullptr) {}
  ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
  ListNode *addTwoNumbers(ListNode *l1, ListNode *l2) {
    // l1->val <= 10, but l2->val < 10
    ListNode *l3;
    ListNode *l1_next;
    if (l1 != nullptr) {
      l3 = l2 != nullptr ? new ListNode(l1->val + l2->val) : l1;
      if (l3->val < 10) {
        l1_next = l1->next;
      } else {
        l3->val -= 10;
        l1_next = l1->next != nullptr
                      ? new ListNode(l1->next->val + 1, l1->next->next)
                      : new ListNode(1);
      }
    } else {
      l1_next = nullptr;
      if (l2 != nullptr) {
        l3 = l2;
      } else {
        return nullptr;
      }
    }

    l3->next = this->addTwoNumbers(l1_next, l2 != nullptr ? l2->next : nullptr);
    return l3;
  }
};

void show_ll(ListNode *ll) {
  if (ll == nullptr) {
    printf("Null");
  } else {
    while (ll->next != nullptr) {
      printf("%d", ll->val);
      ll = ll->next;
    }
    printf("%d\n", ll->val);
  }
}

int main() {
  Solution demo = Solution();
  ListNode *l1 = new ListNode(2, new ListNode(4, new ListNode(3)));
  ListNode *l2 = new ListNode(5, new ListNode(6, new ListNode(4)));
  show_ll(demo.addTwoNumbers(l1, l2));
}